README
LXPanel plugin to connect to wired and wireless networks.


How to build
------------

1. Install dependencies

The dependencies of any Debian project are listed in the "Build-Depends" section
of the file named "control" in the "debian" subdirectory of the project.

If the project has already been released into apt, then the build dependencies
can be automatically installed using the command "sudo apt build-dep <package-name>".

2. Prepare project

To create the "configure" file and prepare the project directory, use the command
"./autogen.sh" in the top directory of the project.

If this file is not present, then instead use the command "autoreconf -i -f"
followed by the command "intltoolize -c --automake --force", both in the top directory
of the project.

3. Configure

To configure the make system, use the command "./configure" in the top directory of
the project. This will by default set the project for installation in the /usr/local tree,
which will not overwrite a version which has been installed from apt.

If you wish to overwrite a preinstalled version in the /usr tree, supply the arguments
"./configure --prefix=/usr --libdir=/usr/lib/<library-location>" to the configure command.
On a 32-bit system, <library-location> should be "arm-linux-gnueabihf".
On a 64-bit system, <library-location> should be "aarch64-linux-gnu".

4. Build

To build the application, use the command "make" in the top directory of the project.

5. Install

To install the application and all required data files, use the command "sudo make install"
in the top directory of the project.


debian/changelog
lxplug-network (0.33) bullseye; urgency=medium

  * Add Korean translations

 -- Simon Long <simon@raspberrypi.com>  Mon, 22 May 2023 08:06:16 +0100

lxplug-network (0.32) bullseye; urgency=medium

  * Faster service check at start

 -- Simon Long <simon@raspberrypi.com>  Wed, 21 Sep 2022 13:53:02 +0100

lxplug-network (0.31) bullseye; urgency=medium

  * Fix for crash when accessing preferences when hidden
  * Add control message to launch menu

 -- Simon Long <simon@raspberrypi.com>  Fri, 15 Jul 2022 11:21:14 +0100

lxplug-network (0.30) bullseye; urgency=medium

  * Check for dhcpcd service before loading plugin

 -- Simon Long <simon@raspberrypi.com>  Wed, 22 Jun 2022 07:39:25 +0100

lxplug-network (0.29) bullseye; urgency=medium

  * Remove scan results message

 -- Simon Long <simon@raspberrypi.com>  Fri, 19 Nov 2021 16:41:48 +0000

lxplug-network (0.28) bullseye; urgency=medium

  * Use standard modal UI for disconnect prompt

 -- Simon Long <simon@raspberrypi.com>  Tue, 29 Jun 2021 15:30:01 +0100

lxplug-network (0.27) bullseye; urgency=medium

  * Fix some memory leaks

 -- Simon Long <simon@raspberrypi.com>  Fri, 21 May 2021 15:33:23 +0100

lxplug-network (0.26) bullseye; urgency=medium

  * Added Armenian translation

 -- Simon Long <simon@raspberrypi.com>  Mon, 26 Apr 2021 08:48:48 +0100

lxplug-network (0.25) bullseye; urgency=medium

  * GTK+3 version

 -- Simon Long <simon@raspberrypi.com>  Fri, 26 Mar 2021 13:27:14 +0000

lxplug-network (0.24) buster; urgency=medium

  * Added Icelandic translation

 -- Simon Long <simon@raspberrypi.com>  Thu, 18 Mar 2021 17:48:29 +0000

lxplug-network (0.23) buster; urgency=medium

  * Added Slovak translation

 -- Simon Long <simon@raspberrypi.com>  Sun, 03 Jan 2021 21:53:47 +0000

lxplug-network (0.22) buster; urgency=medium

  * Norwegian translation added

 -- Simon Long <simon@raspberrypi.com>  Mon, 05 Oct 2020 10:50:15 +0100

lxplug-network (0.21) buster; urgency=medium

  * Italian translation added

 -- Simon Long <simon@raspberrypi.com>  Wed, 15 Jul 2020 08:00:48 +0100

lxplug-network (0.20) buster; urgency=medium

  * Use generic icon load code

 -- Simon Long <simon@raspberrypi.com>  Thu, 16 Apr 2020 15:49:32 +0100

lxplug-network (0.19) buster; urgency=medium

  * Use size_t for all getline calls

 -- Simon Long <simon@raspberrypi.org>  Fri, 28 Feb 2020 07:04:41 +0000

lxplug-network (0.18) buster; urgency=medium

  * Additional fix in octal parser

 -- Simon Long <simon@raspberrypi.org>  Thu, 19 Sep 2019 14:14:15 +0100

lxplug-network (0.17) buster; urgency=medium

  * Fix typo in string decoding

 -- Simon Long <simon@raspberrypi.org>  Tue, 17 Sep 2019 15:39:04 +0100

lxplug-network (0.16) unstable; urgency=medium

  * Use raspi-config to check whether wifi country is set

 -- Simon Long <simon@raspberrypi.org>  Thu, 02 May 2019 10:36:29 +0100

lxplug-network (0.15) stretch; urgency=medium

  * Add config parameters to control background scan frequency

 -- Simon Long <simon@raspberrypi.org>  Tue, 23 Apr 2019 08:22:55 +0100

lxplug-network (0.14) stretch; urgency=medium

  * Check for 5G hardware by querying iw info
  * Don't read back commented-out passwords into network selection box

 -- Simon Long <simon@raspberrypi.org>  Wed, 27 Mar 2019 11:49:02 +0000

lxplug-network (0.13) stretch; urgency=medium

  * Prioritise check for hardware not found over country setting in menu

 -- Simon Long <simon@raspberrypi.org>  Mon, 28 Jan 2019 10:54:26 +0000

lxplug-network (0.12) stretch; urgency=medium

  * Hide PSK characters by default, with check box to toggle

 -- Simon Long <simon@raspberrypi.org>  Thu, 12 Jul 2018 15:07:45 +0100

lxplug-network (0.11) stretch; urgency=medium

  * Recall stored PSK when reconnecting to a network
  * Cosmetic changes to disconnect prompt dialog box

 -- Simon Long <simon@raspberrypi.org>  Thu, 17 May 2018 15:21:18 +0100

lxplug-network (0.10) stretch; urgency=medium

  * Add link to rc_gui to set Wi-Fi country from within plugin menu
  * Bug fix to hide networks with hidden SSIDs correctly

 -- Simon Long <simon@raspberrypi.org>  Mon, 09 Apr 2018 11:08:22 +0100

lxplug-network (0.9) stretch; urgency=medium

  * Merge in recent changes from dhcpcd-ui

 -- Simon Long <simon@raspberrypi.org>  Fri, 02 Mar 2018 07:56:44 +0000

lxplug-network (0.8) stretch; urgency=medium

  * Indicate 5G devices in wi-fi menu

 -- Simon Long <simon@raspberrypi.org>  Wed, 28 Feb 2018 14:22:55 +0000

lxplug-network (0.7) stretch; urgency=medium

  * Warn if 5G device is used with no country set

 -- Simon Long <simon@raspberrypi.org>  Mon, 26 Feb 2018 11:10:40 +0000

lxplug-network (0.6) stretch; urgency=medium

  * Allow devices which are not recognised by rfkill to be controlled

 -- Simon Long <simon@raspberrypi.org>  Fri, 24 Nov 2017 11:32:53 +0000

lxplug-network (0.5) stretch; urgency=medium

  * Fix internationalisation

 -- Simon Long <simon@raspberrypi.org>  Tue, 19 Sep 2017 08:24:22 +0000

lxplug-network (0.4) stretch; urgency=medium

  * Change noipv6rs option to noipv6 option

 -- Simon Long <simon@raspberrypi.org>  Mon, 18 Sep 2017 10:27:46 +0100

lxplug-network (0.3) stretch; urgency=medium

  * Modify preferences dialog for compatibility with IPv6

 -- Simon Long <simon@raspberrypi.org>  Thu, 14 Sep 2017 15:01:05 +0100

lxplug-network (0.2) stretch; urgency=medium

  * Remove unnecessary sudo from rfkill commands

 -- Simon Long <simon@raspberrypi.org>  Tue, 04 Jul 2017 10:20:56 +0100

lxplug-network (0.1) stretch; urgency=medium

  * Initial release.

 -- Simon Long <simon@raspberrypi.org>  Mon, 06 Mar 2017 17:16:23 +0000


plugins/libdhcpcd/dhcpcd.h
/*
 * libdhcpcd
 * Copyright 2009-2015 Roy Marples <roy@marples.name>
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

#ifndef DHCPCD_H
#define DHCPCD_H

#include <net/if.h>
#include <netinet/in.h>

#include <stdbool.h>

#ifdef __cplusplus
extern "C" {
#endif

#ifndef DHCPCD_SOCKET
#define DHCPCD_SOCKET		"/var/run/dhcpcd.sock"
#endif
#ifndef DHCPCD_UNPRIV_SOCKET
#define DHCPCD_UNPRIV_SOCKET	"/var/run/dhcpcd.unpriv.sock"
#endif

#ifndef WPA_CTRL_DIR
#define WPA_CTRL_DIR		"/var/run/wpa_supplicant"
#endif

#ifndef DHCPCD_TMP_DIR
#define DHCPCD_TMP_DIR		"/tmp/dhcpcd"
#endif
#ifndef DHCPCD_TMP_DIR_PERM
#define DHCPCD_TMP_DIR_PERM	0700
#endif

#define DHCPCD_RETRYOPEN	1000	/* milliseconds */
#define DHCPCD_WPA_PING		500	/* milliseconds */
#define DHCPCD_WPA_SCAN_LONG	60000	/* milliseconds */
#define DHCPCD_WPA_SCAN_SHORT	5000	/* milliseconds */
#define DHCPCD_WI_HIST_MAX	10	/* Recall 10 scans for averages */

/* Each non printable byte of the SSID is represented as \000 */
#define IF_SSIDSIZE		((32 * 4) + 1)
#define IF_BSSIDSIZE		64
#define FLAGSIZE		64
#define TYPESIZE		8
#define REASONSIZE		16

#define WPA_FREQ_IS_2G(f)	((f) >= 2402 && (f) <= 2472)
#define WPA_FREQ_IS_5G(f)	((f) >= 5170 && (f) <= 5835)

#define DHC_UNKNOWN		 0
#define DHC_DOWN		 1
#define DHC_OPENED		 2
#define DHC_INITIALISED		 3
#define DHC_DISCONNECTED	 4
#define DHC_CONNECTING		 5
#define DHC_AF_WAITING		 6
#define DHC_CONNECTED		 7
#define DHC_MAX			 8
extern const char * const dhcpcd_cstates[];

#define DHT_UNKNOWN		 0
#define DHT_LINK		 1
#define DHT_DHCP		 2
#define DHT_IPV4		 2
#define DHT_IPV4LL		 3
#define DHT_IPV6		 4
#define DHT_RA			 5
#define DHT_DHCP6		 6
#define DHT_MAX			 7

#define DHS_UNKNOWN		 0
#define DHS_DUMP		 1
#define DHS_TEST		 2
#define DHS_STOPPED		 3
#define DHS_FAIL		 4
#define DHS_STOP		 5
#define DHS_PREINIT		 6
#define DHS_DEPARTED		 7
#define DHS_NOCARRIER		 8
#define DHS_NAK			 9
#define DHS_EXPIRE		10
#define DHS_RECONFIGURE		11
#define DHS_CARRIER		12
#define DHS_STATIC		13
#define DHS_3RDPARTY		14
#define DHS_IPV4LL		15
#define DHS_INFORM		16
#define DHS_BOUND		17
#define DHS_RENEW		18
#define DHS_REBIND		19
#define	DHS_REBOOT		20
#define DHS_ROUTERADVERT	21
#define DHS_DELEGATED		22
#define DHR_MAX			23

typedef struct dhcpcd_wi_avs {
	int value;
	int average;
} DHCPCD_WI_AV;

typedef struct dhcpcd_wi_scan {
	struct dhcpcd_wi_scan *next;
	char bssid[IF_BSSIDSIZE];
	unsigned int flags;
#define WSF_SECURE		0x001
#define WSF_PSK			0x002
#define WSF_WEP			0x010
#define WSF_WPA			0x020
#define WSF_2G			0x100
#define WSF_5G			0x200
	int frequency;
	DHCPCD_WI_AV quality;
	DHCPCD_WI_AV noise;
	DHCPCD_WI_AV level;
	DHCPCD_WI_AV strength;
	char ssid[IF_SSIDSIZE];
	char wpa_flags[FLAGSIZE];
} DHCPCD_WI_SCAN;

#ifdef IN_LIBDHCPCD
typedef struct dhcpcd_if {
	struct dhcpcd_if *next;
	const char *ifname;
	unsigned int type;
	unsigned int state;
	const char *reason;

	unsigned int ifflags;
	bool up;
	bool wireless;
	const char *ssid;
	int freq;

	char *data;
	size_t data_len;

	char *last_message;

	struct dhcpcd_connection *con;
} DHCPCD_IF;
#else
typedef struct dhcpcd_if {
	struct dhcpcd_if *next;
	const char *ifname;
	unsigned int type;
	unsigned int state;
	const char *reason;

	unsigned int ifflags;
	bool up;
	bool wireless;
	const char *ssid;
	int freq;
} DHCPCD_IF;
#endif

typedef struct dhcpcd_config {
	struct dhcpcd_config *next;
	char *option;
	char *value;
} DHCPCD_OPTION;

#ifdef IN_LIBDHCPCD
typedef struct dhcpcd_wi_hist {
	struct dhcpcd_wi_hist *next;
	char ifname[IF_NAMESIZE];
	char bssid[IF_BSSIDSIZE];
	int quality;
	int noise;
	int level;
	int strength;
} DHCPCD_WI_HIST;

typedef struct dhcpcd_wpa {
	struct dhcpcd_wpa *next;
	char ifname[IF_NAMESIZE];
	char userconnect_ssid[IF_SSIDSIZE];
	char connect_ssid[IF_SSIDSIZE];
	unsigned int status;
	int command_fd;
	char *command_path;
	int listen_fd;
	char *listen_path;
	bool attached;
	struct dhcpcd_connection *con;
	void *context;
} DHCPCD_WPA;

typedef struct dhcpcd_connection {
	struct dhcpcd_connection *next;
	bool open;
	bool privileged;
	int command_fd;
	int listen_fd;
	const char *progname;

	DHCPCD_IF *interfaces;
	DHCPCD_WPA *wpa;
	DHCPCD_WI_HIST *wi_history;

	void (*if_cb)(DHCPCD_IF *, void *);
	void *if_context;
	void (*status_cb)(struct dhcpcd_connection *,
	    unsigned int, const char *, void *);
	void *status_context;
	bool wpa_started;
	void (*wi_scanresults_cb)(DHCPCD_WPA *, void *);
	void *wi_scanresults_context;
	void (*wi_error_cb)(DHCPCD_WPA *, void *);
	void *wi_error_context;
	void (*wpa_status_cb)(DHCPCD_WPA *, unsigned int, const char *, void *);
	void *wpa_status_context;

	char *buf;
	size_t buflen;

	char *version;
	bool terminate_commands;
	char *error;
	int err;
	int errors;
	unsigned int status;
	bool af_waiting;

	char *cffile;
} DHCPCD_CONNECTION;

#else
typedef void *DHCPCD_CONFIG;
typedef void *DHCPCD_WPA;
typedef void *DHCPCD_CONNECTION;
#endif

DHCPCD_CONNECTION * dhcpcd_new(void);
const char * dhcpcd_version(DHCPCD_CONNECTION *);
void dhcpcd_set_progname(DHCPCD_CONNECTION *, const char *);
const char * dhcpcd_get_progname(const DHCPCD_CONNECTION *);
unsigned int dhcpcd_status(DHCPCD_CONNECTION *, const char **);
bool dhcpcd_af_waiting(const DHCPCD_CONNECTION *);
const char * dhcpcd_cffile(DHCPCD_CONNECTION *);
bool dhcpcd_realloc(DHCPCD_CONNECTION *, size_t);
int dhcpcd_open(DHCPCD_CONNECTION *, bool priv);
void dhcpcd_close(DHCPCD_CONNECTION *);
void dhcpcd_free(DHCPCD_CONNECTION *);
void dhcpcd_set_if_callback(DHCPCD_CONNECTION *,
    void (*)(DHCPCD_IF *, void *), void *);
void dhcpcd_set_status_callback(DHCPCD_CONNECTION *,
    void (*)(DHCPCD_CONNECTION *, unsigned int, const char *, void *), void *);
int dhcpcd_get_fd(DHCPCD_CONNECTION *);
bool dhcpcd_privileged(DHCPCD_CONNECTION *);
void dhcpcd_dispatch(DHCPCD_CONNECTION *);
DHCPCD_IF * dhcpcd_interfaces(DHCPCD_CONNECTION *);
char **dhcpcd_interface_names(DHCPCD_CONNECTION *, size_t *);
void dhcpcd_freev(char **);
char **dhcpcd_interface_names_sorted(DHCPCD_CONNECTION *);
DHCPCD_IF * dhcpcd_get_if(DHCPCD_CONNECTION *, const char *, unsigned int);
DHCPCD_CONNECTION * dhcpcd_if_connection(DHCPCD_IF *);
const char *dhcpcd_get_value(const DHCPCD_IF *, const char *);
const char *dhcpcd_get_prefix_value(const DHCPCD_IF *, const char *,
    const char *);
ssize_t dhcpcd_encode_string_escape(char *, size_t, const char *, size_t);
ssize_t dhcpcd_decode_string_escape(char *, size_t, const char *);
ssize_t dhcpcd_decode_hex(char *, size_t, const char *);
char * dhcpcd_if_message(DHCPCD_IF *, bool *);

ssize_t dhcpcd_command(DHCPCD_CONNECTION *, const char *, char **);
ssize_t dhcpcd_command_arg(DHCPCD_CONNECTION *, const char *, const char *,
    char **);
#define dhcpcd_rebind(c, i)	dhcpcd_command_arg((c), "-n", (i), NULL)
#define dhcpcd_release(c, i)	dhcpcd_command_arg((c), "-k", (i), NULL)

void dhcpcd_wpa_start(DHCPCD_CONNECTION *);
DHCPCD_WPA *dhcpcd_wpa_find(DHCPCD_CONNECTION *, const char *);
DHCPCD_WPA *dhcpcd_wpa_new(DHCPCD_CONNECTION *, const char *);
DHCPCD_CONNECTION *dhcpcd_wpa_connection(DHCPCD_WPA *);
int dhcpcd_wpa_open(DHCPCD_WPA *);
void dhcpcd_wpa_close(DHCPCD_WPA *);
void dhcpcd_wpa_dispatch(DHCPCD_WPA *);
int dhcpcd_wpa_get_fd(DHCPCD_WPA *);
DHCPCD_IF *dhcpcd_wpa_if(DHCPCD_WPA *);
void dhcpcd_wpa_if_event(DHCPCD_IF *);
void dhcpcd_wpa_set_scan_callback(DHCPCD_CONNECTION *,
    void (*)(DHCPCD_WPA *, void *), void *);
 void
dhcpcd_wpa_set_error_callback(DHCPCD_CONNECTION *con,
    void (*cb)(DHCPCD_WPA *, void *), void *context);

void dhcpcd_wpa_set_status_callback(DHCPCD_CONNECTION *,
    void (*)(DHCPCD_WPA *, unsigned int, const char *, void *), void *);
int dhcpcd_wi_scan_compare(DHCPCD_WI_SCAN *a, DHCPCD_WI_SCAN *b);
DHCPCD_WI_SCAN * dhcpcd_wi_scans(DHCPCD_IF *);
bool dhcpcd_wi_associated(DHCPCD_IF *i, DHCPCD_WI_SCAN *s);
void dhcpcd_wi_scans_free(DHCPCD_WI_SCAN *);
void dhcpcd_wi_history_clear(DHCPCD_CONNECTION *);
bool dhcpcd_wpa_set_network(DHCPCD_WPA *, int, const char *, const char *);
int dhcpcd_wpa_find_network_new(DHCPCD_WPA *, const char *);
bool dhcpcd_wpa_command(DHCPCD_WPA *, const char *);
bool dhcpcd_wpa_command_arg(DHCPCD_WPA *, const char *, const char *);
unsigned int dhcpcd_wpa_status(DHCPCD_WPA *, const char **);
int dhcpcd_wpa_freq(DHCPCD_WPA *);
#define WST_BSSID	0x01
#define WST_FLAGS	0x02
#define WST_FREQ	0x03
int dhcpcd_wi_print_tooltip(char *, size_t, DHCPCD_WI_SCAN *, unsigned int);

bool dhcpcd_wpa_ping(DHCPCD_WPA *);
bool dhcpcd_wpa_can_background_scan(DHCPCD_WPA *);
bool dhcpcd_wpa_scan(DHCPCD_WPA *);
bool dhcpcd_wpa_reconfigure(DHCPCD_WPA *);
bool dhcpcd_wpa_reassociate(DHCPCD_WPA *);
bool dhcpcd_wpa_disconnect(DHCPCD_WPA *);
bool dhcpcd_wpa_config_write(DHCPCD_WPA *);
int dhcpcd_wpa_network_find_new(DHCPCD_WPA *, const char *);
bool dhcpcd_wpa_network_disable(DHCPCD_WPA *, int);
bool dhcpcd_wpa_network_enable(DHCPCD_WPA *, int);
bool dhcpcd_wpa_network_select(DHCPCD_WPA *, int);
bool dhcpcd_wpa_network_remove(DHCPCD_WPA *, int);
char * dhcpcd_wpa_network_get(DHCPCD_WPA *, int, const char *);
bool dhcpcd_wpa_network_set(DHCPCD_WPA *, int, const char *, const char *);

#define DHCPCD_WPA_SUCCESS	 0
#define DHCPCD_WPA_ERR		-1
#define DHCPCD_WPA_ERR_SET	-2
#define DHCPCD_WPA_ERR_SET_PSK	-3
#define DHCPCD_WPA_ERR_ENABLE	-4
#define DHCPCD_WPA_ERR_WRITE	-5
#define DHCPCD_WPA_ERR_SELECT	-6
#define DHCPCD_WPA_ERR_ASSOC	-7
#define DHCPCD_WPA_ERR_DISCONN	-8
#define DHCPCD_WPA_ERR_RECONF	-9
int dhcpcd_wpa_configure(DHCPCD_WPA *w, DHCPCD_WI_SCAN *s, const char *p);
int dhcpcd_wpa_select(DHCPCD_WPA *w, DHCPCD_WI_SCAN *s);

void dhcpcd_wpa_set_context (DHCPCD_WPA *wpa, void *context);
void *dhcpcd_wpa_get_context (DHCPCD_WPA *wpa);

char ** dhcpcd_config_blocks(DHCPCD_CONNECTION *, const char *);
DHCPCD_OPTION *dhcpcd_config_read(DHCPCD_CONNECTION *,
    const char *, const char *);
void dhcpcd_config_free(DHCPCD_OPTION *);
const char * dhcpcd_config_get(DHCPCD_OPTION *, const char *);
const char * dhcpcd_config_get_static(DHCPCD_OPTION *, const char *);
bool dhcpcd_config_set(DHCPCD_OPTION **, const char *, const char *);
bool dhcpcd_config_set_static(DHCPCD_OPTION **, const char *, const char *);
bool dhcpcd_config_writeable(DHCPCD_CONNECTION *con);
bool dhcpcd_config_write(DHCPCD_CONNECTION *,
    const char *, const char *, const DHCPCD_OPTION *);

bool dhcpcd_is_wireless (DHCPCD_IF *i);

#ifdef __cplusplus
}
#endif

#endif
